src/server/src/main/java/io/cassandrareaper/resources/RepairRunResource.java

```diff
@@ -381,6 +393,23 @@ public Response modifyRunState(
     }
     final RepairRun.RunState oldState = repairRun.get().getRunState();
 
+    final boolean isIntensityPresent = intensityStr.isPresent();
+    if (isIntensityPresent && (oldState != RepairRun.RunState.PAUSED || newState != RepairRun.RunState.PAUSED)) {
+      return Response.status(Response.Status.BAD_REQUEST.getStatusCode())
+              .entity("intensity can be modified when repair run already paused")
+              .build();
+    }
+
+    if (isIntensityPresent) {
+      final Response intensityValidationResponse = validateIntensity(intensityStr);
+      if (intensityValidationResponse != null) {
+        return intensityValidationResponse;
+      }
+
+      final double intensity = Double.parseDouble(intensityStr.get());
+      return editRun(repairRun.get(), repairUnit.get(), segmentsRepaired, intensity);
```

Comment:
Continued from https://github.com/thelastpickle/cassandra-reaper/pull/303#discussion_r157944766

> So there will be just one PUT method but it will reroute call through finite state machine graph.

That would not be idempotent, which is required of PUT requests. (And is desired for of our all endpoints.)

> I thought about another endpoint but current logic means…

The existing endpoint isn't perfect, but is difficult to remove.
So I'd suggest adding the two following endpoints:

 - `PUT repair_run/{id}/state/{state}`
 - `PUT repair_run/{id}/intensity/{intensity}`

And change the existing endpoint `PUT /repair_run/{id}?state={state}` return a MOVED_PERMANENT to `PUT repair_run/{id}/state/{state}`.

Then logic for each endpoint would be:

 - `PUT /repair_run/{id}?state={state}`

 1. if request does not specify new state, then BAD_REQUEST is returned,
 1. if the repair run (or unit) is not found, then NOT_FOUND is returned,
 1. return MOVED_PERMANENTLY to `PUT repair_run/{id}/state/{state}`.

 - `PUT repair_run/{id}/state/{state}`

 1. if request does not specify new state, then BAD_REQUEST is returned,
 1. if the repair run (or unit) is not found, then NOT_FOUND is returned,
 1. if there's any other repair runs for the related repair unit and they are RUNNING, then return CONFLICT,
 1. if state is unaltered, all other parameters are ignored, and NOT_MODIFIED is returned,
 1. if state is altered, but an invalid transition, then all other parameters are ignored, and METHOD_NOT_ALLOWED returned,
 1. if state is altered, the transition valid, then all other parameters are ignored, and OK returned.

 - `PUT repair_run/{id}/intensity/{intensity}`

 1. if request does not specify new state, then BAD_REQUEST is returned,
 1. if the repair run (or unit) is not found, then NOT_FOUND is returned,
 1. if there's any other repair runs for the related repair unit and they are RUNNING, then return CONFLICT,
 1. if state is not PAUSED then return CONFLICT,
 1. if state is PAUSED, intensity specified but invalid, then BAD_REQUEST is returned,
 1. if state is PAUSED, intensity specified and valid, then OK is returned,

Reply:
I've made some refactoring. But it still looks like spaghetti code. You should look to Railway Oriented Programming for validation: https://fsharpforfunandprofit.com/rop/
There are examples for C#/F# but I hope you can understand concept.

Reply:
Take a look at the rewrite i did in https://github.com/thelastpickle/cassandra-reaper/commit/e3e772945f1b8676140a74a944fbb3fc1a33e964

If we wanted to separate the validation concerns the approach for this code would be via jaxrs validation:  https://samaxes.com/2014/04/jaxrs-beanvalidation-javaee7-wildfly/

Reply:
Actually I tired of arguing about this PR for a week. I just want to let it go. If you think that it's okay, accept PR. I just need possibility to change intensity.

Reply:
I read you changes. I understand what you mean. Would you like to merge it to master?

Reply:
> I read your changes and understand what you mean. Would you like to merge it to master?

Ok. I will merge it. Thanks.

> Actually I tired of arguing about this PR for a week. I just want to let it go. 

Please don't be. Putting the extra effort into each contribution is always worth it, especially as it's often someone else that has to fix or build something on top of it later on.
Again, thanks for the contribution and effort.